File: bank-service/src/main/java/rs/edu/raf/controller/BankAccountController.java
package rs.edu.raf.controller;

import io.swagger.annotations.ApiOperation;
import io.swagger.v3.oas.annotations.Parameter;
import io.swagger.v3.oas.annotations.security.SecurityRequirement;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import lombok.AllArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import rs.edu.raf.dto.BankAccountDTO;
import rs.edu.raf.dto.BusinessBankAccountCreateDTO;
import rs.edu.raf.dto.CurrentBankAccountCreateDTO;
import rs.edu.raf.dto.ForeignCurrencyBankAccountCreateDTO;
import rs.edu.raf.security.JwtUtil;
import rs.edu.raf.service.BankAccountService;

import java.math.BigDecimal;
import java.util.List;

@RestController
@RequestMapping("/bank-accounts")
@Tag(name = "Bank accounts", description = "Managing bank accounts")
@AllArgsConstructor
@SecurityRequirement(name = "jwt")
public class BankAccountController {
    private BankAccountService bankAccountService;
    private JwtUtil jwtUtil;

    @ApiOperation(value = "Create foreign currency bank account")
    @PostMapping("/foreign-currency")
    public ResponseEntity<BankAccountDTO> createForeignCurrencyBankAccount(
            @RequestBody @Valid ForeignCurrencyBankAccountCreateDTO foreignCurrencyBankAccountCreateDTO) {
        return new ResponseEntity<>(
                bankAccountService.createForeignCurrencyBankAccount(foreignCurrencyBankAccountCreateDTO,
                jwtUtil.getRealIDForLoggedUser()), HttpStatus.OK
        );
    }

    @ApiOperation(value = "Create business bank account")
    @PostMapping("/business")
    public ResponseEntity<BankAccountDTO> createBusinessBankAccount(
            @RequestBody @Valid BusinessBankAccountCreateDTO businessBankAccountCreateDTO){
        return new ResponseEntity<>(bankAccountService.createBusinessBankAccount(businessBankAccountCreateDTO,
                jwtUtil.getRealIDForLoggedUser()), HttpStatus.OK);
    }
    @ApiOperation(value = "Create current bank account")
    @PostMapping("/current")
    public ResponseEntity<BankAccountDTO> createCurrentBankAccount(
            @RequestBody @Valid CurrentBankAccountCreateDTO currentBankAccountCreateDTO){
        return new ResponseEntity<>(bankAccountService.createCurrentBankAccount(currentBankAccountCreateDTO,
                jwtUtil.getRealIDForLoggedUser()), HttpStatus.OK);
    }
    @ApiOperation(value = "Get all bank accounts for client")
    @GetMapping("/client/{id}")
    public ResponseEntity<List<BankAccountDTO>> getAllBankAccountsForClient(@Parameter(name = "Client ID") @PathVariable("id") Long id){
        return new ResponseEntity<>(bankAccountService.getAllBankAccountsForOwner(id), HttpStatus.OK);
    }

    @ApiOperation(value = "Get all bank account for logged client")
    @GetMapping("/client")
    public ResponseEntity<List<BankAccountDTO>> getAllBankAccountsForLoggedClient(){
        return new ResponseEntity<>(bankAccountService.getAllBankAccountsForOwner(jwtUtil.getIDForLoggedUser()), HttpStatus.OK);
    }

    @ApiOperation(value = "Get bank account with ID")
    @GetMapping("/account/{id}")
    public ResponseEntity<BankAccountDTO> getBankAccountWithID(@Parameter(name = "Bank account ID") @PathVariable("id") Long id){
        return new ResponseEntity<>(bankAccountService.findActiveBankAccountByID(id), HttpStatus.OK);
    }

    @ApiOperation(value = "Get bank account by account number")
    @GetMapping("/account/number/{accountNumber}")
    public ResponseEntity<BankAccountDTO> getBankAccountWithAccountNumber(@Parameter(name = "Bank account number")
                                                              @PathVariable("accountNumber") Long bankAccountNumber) {
        return new ResponseEntity<>(bankAccountService.findActiveBankAccountByAccountNumber(bankAccountNumber), HttpStatus.OK);
    }

    @ApiOperation(value = "Deactivate bank account")
    @PutMapping("/deactivate/{accountNumber}")
    public ResponseEntity<Boolean> deactivateBankAccount(@Parameter(name = "Bank account number")
                                                             @PathVariable("accountNumber") Long accountNumber) {
        return new ResponseEntity<>(bankAccountService.deactivateBankAccount(accountNumber),HttpStatus.OK);
    }

    @ApiOperation(value = "Deducts available balance")
    @PutMapping("/deduct-available-balance/{owner}/{amount}")
    public ResponseEntity<Boolean> deductAvailableBalance(@Parameter(name = "Amount") @PathVariable("amount") BigDecimal amount,
            @Parameter(name = "Bank account owner ID") @PathVariable("owner") Long accountNumber) {
        return new ResponseEntity<>(bankAccountService.deductAvailableBalanceFromBankAccount(accountNumber, amount), HttpStatus.OK);
    }
}

File: bank-service-kotlin/src/main/kotlin/org/example/bankservicekotlin/controller/BankAccountController.kt
package org.example.bankservicekotlin.controller

import jakarta.validation.Valid
import org.example.bankservicekotlin.dto.BankAccountDTO
import org.example.bankservicekotlin.dto.BusinessBankAccountCreateDTO
import org.example.bankservicekotlin.dto.CurrentBankAccountCreateDTO
import org.example.bankservicekotlin.dto.ForeignCurrencyBankAccountCreateDTO
import org.example.bankservicekotlin.security.JwtUtil
import org.example.bankservicekotlin.service.BankAccountService
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.math.BigDecimal

@RestController
@RequestMapping("/kotlin/bank-accounts")

class BankAccountController(
    private val bankAccountService: BankAccountService,
    private val jwtUtil: JwtUtil) {

    @PostMapping("/foreign-currency")
    fun createForeignCurrencyBankAccount(
        @RequestBody foreignCurrencyBankAccountCreateDTO: @Valid ForeignCurrencyBankAccountCreateDTO
    ): ResponseEntity<BankAccountDTO> {
        return ResponseEntity<BankAccountDTO>(
            bankAccountService.createForeignCurrencyBankAccount(
                foreignCurrencyBankAccountCreateDTO,
                jwtUtil.getRealIDForLoggedUser()
            ), HttpStatus.OK
        )
    }

    @PostMapping("/business")
    fun createBusinessBankAccount(
        @RequestBody businessBankAccountCreateDTO: @Valid BusinessBankAccountCreateDTO
    ): ResponseEntity<BankAccountDTO> {
        return ResponseEntity<BankAccountDTO>(
            bankAccountService.createBusinessBankAccount(
                businessBankAccountCreateDTO,
                jwtUtil.getRealIDForLoggedUser()
            ), HttpStatus.OK
        )
    }

    @PostMapping("/current")
    fun createCurrentBankAccount(
        @RequestBody currentBankAccountCreateDTO: @Valid CurrentBankAccountCreateDTO
    ): ResponseEntity<BankAccountDTO> {
        return ResponseEntity<BankAccountDTO>(
            bankAccountService.createCurrentBankAccount(
                currentBankAccountCreateDTO,
                jwtUtil.getRealIDForLoggedUser()
            ), HttpStatus.OK
        )
    }

    @GetMapping("/client/{id}")
    fun getAllBankAccountsForClient(@PathVariable("id") id: Long): ResponseEntity<List<BankAccountDTO>> {
        return ResponseEntity<List<BankAccountDTO>>(bankAccountService.getAllBankAccountsForOwner(id), HttpStatus.OK)
    }

    @GetMapping("/client/currency/{currency}")
    fun getBankAccountsWithCurrencyForLoggedClient(
        @PathVariable("currency") id: Long
    ): ResponseEntity<List<BankAccountDTO>> {
        return ResponseEntity<List<BankAccountDTO>>(
            bankAccountService.getAllBankAccountsWithSpecifiedCurrencyForOwner(
                jwtUtil.getIDForLoggedUser(),
                id
            ), HttpStatus.OK
        )
    }

    @GetMapping("/client")
    fun getAllBankAccountsForLoggedClient(): ResponseEntity<List<BankAccountDTO>> {
        return ResponseEntity<List<BankAccountDTO>>(
            bankAccountService.getAllBankAccountsForOwner(jwtUtil.getIDForLoggedUser()),
            HttpStatus.OK
        )
    }

    @GetMapping
    fun getAllBankAccounts(): ResponseEntity<List<BankAccountDTO>> {
        return ResponseEntity<List<BankAccountDTO>>(bankAccountService.getAllBankAccounts(), HttpStatus.OK)
    }

    @GetMapping("/account/{id}")
    fun getBankAccountWithID(@PathVariable("id") id: Long): ResponseEntity<BankAccountDTO> {
        return ResponseEntity<BankAccountDTO>(bankAccountService.findActiveBankAccountByID(id), HttpStatus.OK)
    }

    @GetMapping("/account/number/{accountNumber}")
    fun getBankAccountWithAccountNumber(@PathVariable("accountNumber") bankAccountNumber: Long): ResponseEntity<BankAccountDTO> {
        return ResponseEntity<BankAccountDTO>(
            bankAccountService.findActiveBankAccountByAccountNumber(bankAccountNumber),
            HttpStatus.OK
        )
    }

    @PutMapping("/deactivate/{accountNumber}")
    fun deactivateBankAccount(@PathVariable("accountNumber") accountNumber: Long): ResponseEntity<Boolean> {
        return ResponseEntity<Boolean>(bankAccountService.deactivateBankAccount(accountNumber), HttpStatus.OK)
    }

    @PutMapping("/deduct-available-balance/{owner}/{amount}")
    fun deductAvailableBalance(
        @PathVariable("amount") amount: BigDecimal,
        @PathVariable("owner") accountNumber: Long
    ): ResponseEntity<Boolean> {
        return ResponseEntity<Boolean>(
            bankAccountService.deductAvailableBalanceFromBankAccount(
                accountNumber,
                amount
            ), HttpStatus.OK
        )
    }
}